openmc/cmfd.py

```diff
@@ -518,3 +603,2717 @@ def export_to_xml(self):
+        for g in range(ng):
+            for h in range(ng):
+                # Extract all regions where a cell is a fuel region
+                condition = self._coremap != _CMFD_NOACCEL
+                idx_x = ng * (self._coremap[condition]) + g
+                idx_y = ng * (self._coremap[condition]) + h
+                # Get nu-fission macro xs, transposed
+                if adjoint:
+                    vals = (self._nfissxs[:, :, :, g, h])[condition]
+                # Get nu-fission macro xs
+                else:
+                    vals = (self._nfissxs[:, :, :, h, g])[condition]
+                # Store rows, cols, and data to add to CSR matrix
+                row = np.append(row, idx_x)
+                col = np.append(col, idx_y)
+                data = np.append(data, vals)
+
+        # Create csr matrix
+        prod = sparse.csr_matrix((data, (row, col)), shape=(n, n))
+        return prod
+
+    def _matrix_to_indices(self, irow, nx, ny, nz, ng):
+        """Converts matrix index in CMFD matrices to spatial and group indices
+        of actual problem, based on values from coremap
+
+        Parameters
+        ----------
+        irow : int
+            Row in CMFD matrix
+        nx : int
+            Total number of mesh cells in problem in x direction
+        ny : int
+            Total number of mesh cells in problem in y direction
+        nz : int
+            Total number of mesh cells in problem in z direction
+        ng : int
+            Total number of energy groups in problem
+
+        Returns
+        -------
+        i : int
+            Corresponding x-index in CMFD problem
+        j : int
+            Corresponding y-index in CMFD problem
+        k : int
+            Corresponding z-index in CMFD problem
+        g : int
+            Corresponding group in CMFD problem
+
+        """
+        g = irow % ng
+        # Get indices from coremap
+        spatial_idx = np.where(self._coremap == int(irow/ng))
+        i = spatial_idx[0][0]
+        j = spatial_idx[1][0]
+        k = spatial_idx[2][0]
+
+        return i, j, k, g
+
+    def _indices_to_matrix(self, i, j, k, g, ng):
+        """Takes (i,j,k,g) indices and computes location in CMFD matrix
+
+        Parameters
+        ----------
+        i : int
+            Corresponding x-index in CMFD problem
+        j : int
+            Corresponding y-index in CMFD problem
+        k : int
+            Corresponding z-index in CMFD problem
+        g : int
+            Corresponding group in CMFD problem
+        ng : int
+            Total number of energy groups in problem
+
+        Returns
+        -------
+        matidx : int
+            Row in CMFD matrix
+
+        """
+        # Get matrix index from coremap
+        matidx = ng*(self._coremap[i,j,k]) + g
+        return matidx
+
+    def _execute_power_iter(self, loss, prod):
+        """Main power iteration routine for the CMFD calculation
+
+        Para
```

Comment:
Agreed that this vectorized code is pretty tough to read.  But I think we can try to make it a bit more readable.  In this case, we can clear things up a bit by making aliases for variables like `self._albedo[0]`, splitting one of these `np.where`s into an `if` block, and using a lot of tuples for indexing:
```python
        where_is_accel = np.where(is_accel)
        boundary = (0, where_is_accel[1], where_is_accel[2])
        boundary_grps = boundary + (slice(None),)
        D = self._diffcof[boundary_grps]
        dz = self._hxyz[boundary + (np.newaxis, 0)]
        if is_zero_flux_alb[0]:
            self._dtilde[boundary_grps + (0,)] = 2.0 * D / dz
        else:
            alb = self._albedo[0]
            self._dtilde[boundary_grps + (0,)] = ((2 * D * (1 - alb))
                 / (4 * D * (1 + alb) + (1 - alb) * dz))
```

Reply:
Ah, whoops.  Something like `mask` is fine, but I'm hoping we can do this without creating any new arrays that are the full 3D size.  Would this or something close to it work?
```python
        nx, ny, nz = self._coremap.shape
        y_inds, z_inds = np.indices((ny, nz))
        boundary_accel = is_accel[(0, y_inds, z_inds)]
        boundary = (0, y_inds[boundary_accel], z_inds[boundary_accel])
```
Also, am I correct in thinking that `is_accel` is fixed and doesn't change over the course of a run?  In that case, we can also precompute these boundary indexing arrays at the start of the run and reuse them at each iteration.

Reply:
This looks much better, thanks again. Also, you are correct, I will work on precomputing the boundary indexing arrays